### PR TITLE
Show how to use Hypothesis to fuzz Python code

### LINKED ISSUE #4121: Fuzzing support for Python projects?
I'm a core developer of Hypothesis, the property-based testing library for Python.  Hypothesis is secretly a structured fuzzer, and [in March I added](https://github.com/HypothesisWorks/hypothesis/pull/2371) support for [treating arbitrary property-based tests as fuzz targets](https://hypothesis.readthedocs.io/en/latest/details.html#use-with-external-fuzzers).

It would be lovely to apply this toolchain to CPython ([see here](https://github.com/Zac-HD/stdlib-property-tests/), [and here](https://pyfound.blogspot.com/2020/05/property-based-testing-for-python.html)), write fuzzable tests for `pip`, fuzz existing tests for foundational tools like Numpy/Mercurial/PyPy, and so on.  The only catch is that it's unclear to me whether, and if so how, Python projects are actually supported in OSS-FUZZ!

For example, to exercise the Python JSON module with `python-afl`:

```python
import json, os, sys
import afl
from hypothesis import given, strategies as st

@given(st.recursive(
    st.none() | st.booleans() | st.integers() | st.floats(allow_nan=False) | st.text(),
    lambda x: st.lists(x) | st.dictionaries(st.text(), x)
))
def test_json_rountrip(x):
    assert x == json.loads(json.dumps(x))

if __name__ == "__main__":
    fuzz_target = test_json_rountrip.hypothesis.fuzz_one_input
    afl.init()
    fuzz_target(sys.stdin.buffer)
    os._exit(0)
```

If upstream support in Hypothesis would make OSS-FUZZ integration easier, I would be very happy to add whatever interface we decide on :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by Zac-HD**
Hypothesis makes fuzzing complex Python code fun, easy, and a lot more rewarding than constructing all your objects by hand.

CC @inferno-chromium; closes #4121.

### CODE REVIEW

File: projects/ujson/build.sh [hunk 1]
```diff
@@ -18,6 +18,9 @@
 # Build and install project (using current CFLAGS, CXXFLAGS).
 pip3 install .
 
+# Install Hypothesis for the structured fuzz harness
+pip3 install hypothesis
```
Comment (catenacyber) [inline projects/ujson/build.sh:22]: Should this be done in the base-builder image ?

Comment (Zac-HD) [reply projects/ujson/build.sh:22]: Maybe! It's a fairly lightweight package (only two transitive dependencies), but that's still a bunch of code which most fuzzers don't need in the container... or the convenience might win out.

I think most Hypothesis-based fuzz harnesses will have other dependencies though, such as Hypothesmith, ao it makes sense to install it explicitly here.

Comment (catenacyber) [reply projects/ujson/build.sh:22]: In this case, maybe it is better in Dockerfile than build.sh, what do you think ?

Comment (inferno-chromium) [reply projects/ujson/build.sh:22]: Yes please move this to Dockerfile. If many projects start using this, then we can move this install to base-builder (similar to Atheris).

File: docs/getting-started/new-project-guide/python_lang.md [hunk 2]
```diff
@@ -27,12 +27,33 @@ Python fuzzing in OSS-Fuzz depends on
 `atheris` package, and dependencies are pre-installed on the OSS-Fuzz base
 docker images.
 
+## Hypothesis
```
Comment (inferno-chromium) [inline docs/getting-started/new-project-guide/python_lang.md:30]: This makes it slightly confusing to put Hypothesis between Atheris description and its required project file changes for Atheris. I think it is better to move this to end, after "Project files" to its own section. THat way, you give Hypothesis description, required change to go in dockerfile, give a simple example where hypothesis value comes in and then at end give hyperlink to ujson fuzzer with hypothesis. wdyt ? 

## Hypothesis - Property-based testing

Have same description

then say, like to use Hypothesis, add it in Dockerfile

```shell
pip3 install hypothesis
```

maybe show a simple example

And then give hyperlink to this ujson example fuzzer, so people can look at actual code of a complicated one

File: projects/ujson/hypothesis_structured_fuzzer.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/python3
+
+# Copyright 2020 Zac Hatfield-Dodds
```
Comment (inferno-chromium) [inline projects/ujson/hypothesis_structured_fuzzer.py:3]: 2021

File: projects/ujson/hypothesis_structured_fuzzer.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (inferno-chromium) [inline projects/ujson/hypothesis_structured_fuzzer.py:5]: @Google-Autofuzz @IanPudney - Do you want to review this code.

**Review by inferno-chromium — CHANGES_REQUESTED** [review]

**Comment by Zac-HD** [comment]
@inferno-chromium done 😁

**Review by inferno-chromium — APPROVED** [review]

**Pull request merged**